File: online_judge/LC/1351_Count_Negative_Numbers_in_a_Sorted_Matrix/solution.py
# solution 1
class Solution:
    def countNegatives(self, grid: List[List[int]]) -> int:
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] < 0:
                    count += 1

        return count

# solution 2
class Solution:
    def countNegatives(self, grid: List[List[int]]) -> int:
        count = 0
        m = len(grid)
        n = len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j] < 0:
                    count += n - j
                    break

        return count
    
# solution 3
class Solution:
    def countNegatives(self, grid: List[List[int]]) -> int:
        count = 0
        n = len(grid[0])
        for i in range(len(grid)):
            left, right, mid = 0, n - 1, 0
            while left <= right:
                mid = (left + right) // 2
                if grid[i][mid] < 0:
                    right = mid - 1
                else:
                    left = mid + 1
            count += n - left

        return count
    
# solution 4
class Solution:
    def countNegatives(self, grid: List[List[int]]) -> int:
        count = 0
        m = len(grid) - 1
        n = 0
        column_size = len(grid[0])
        while m >= 0 and n < column_size:
            if grid[m][n] >= 0:
                n += 1
            else:
                count += column_size - n
                m -= 1

        return count

File: online_judge/LC/1351_Count_Negative_Numbers_in_a_Sorted_Matrix/solution.cpp
// solution 1
class Solution {
public:
    int countNegatives(vector<vector<int>>& grid) {
        int count = 0;
        int m = grid.size();
        int n = grid[0].size();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] < 0) {
                    count++;
                }
            }
        }

        return count;
    }
};

// solution 2
class Solution {
public:
    int countNegatives(vector<vector<int>>& grid) {
        int count = 0;
        int m = grid.size();
        int n = grid[0].size();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] < 0) {
                    count += n - j;
                    break;
                }
            }
        }

        return count;
    }
};

// solution 3
class Solution {
public:
    int countNegatives(vector<vector<int>>& grid) {
        int count = 0;
        int n = grid[0].size();
        for (int i = 0; i < grid.size(); i++) {
            int left = 0;
            int right = n - 1;
            int mid;
            while (left <= right) {
                mid = left + (right - left) / 2;
                if (grid[i][mid] < 0) {
                    right = mid-1;
                } else {
                    left = mid+1;
                }
            }
            count += n - left;
        }
        return count;
    }
};

// solution 4
class Solution {
public:
    int countNegatives(vector<vector<int>>& grid) {
        int count = 0;
        int m = grid.size() - 1;
        int column_size = grid[0].size();
        int n = 0;
        while (m >= 0 && n < column_size) {
            if (grid[m][n] >= 0) {
                n++;
            } else {
                count += column_size - n;
                m--;
            }
        }

        return count;
    }
};